src/torchjd/autojac/_utils.py

```diff
@@ -30,3 +31,75 @@ def _check_retain_graph_compatible_with_chunk_size(
             "When using `retain_graph=False`, parameter `parallel_chunk_size` must be `None` or "
             "large enough to compute all gradients in parallel."
         )
+
+
+def _traverse_ag_graph(grad_graph: Node) -> list[Tensor]:
+    """
+    Traverses an autograd graph recursively and extracts all variables (trainable parameters) that lead to the node.
+    @param grad_graph: autograd computation node
+    @return: found params of calculation
+    """
+    child_nodes = grad_graph.next_functions
+    output_params = []
+    for child in child_nodes:
+        if child[0] is None:
+            # constants in calculation case
+            continue
+        if hasattr(child[0], "variable"):
+            output_params.append(child[0].variable)
+        else:
+            output_params += _traverse_ag_graph(child[0])
+    return output_params
+
+
+def _determine_shared_not_shared(
+    tensor_list: list[list[Tensor]],
+) -> Tuple[list[Tensor], list[list[Tensor]]]:
+    """
+    Based on a list of lists of tensor objects we identify the Tensor objects that are shared in ALL tensor lists
+    and the tensor objects that only appear in a tensor list.
+    @param tensor_list: list of list containing pytorch Tensor objects
+    @return: a tuple containing all shared tensors as a list and a list of list of all individual tensors
+    """
+    count_map = {}
+    for tensors in tensor_list:
+        for memory_address in set(
+            [id(tensor) for tensor in tensors]
+        ):  # we use set to ensure uniqueness
+            if memory_address in count_map:
+                count_map[memory_address] += 1
+            else:
+                count_map[memory_address] = 1
+
+    shared_tensors = []
+    shared_already_appended = set()
+    task_parameters = []
+    search_shared = True
+    for tensors in tensor_list:
```

Comment:
It feels this code could be simplified by using a Counter (https://docs.python.org/3/library/collections.html#collections.Counter)

Reply:
done